reconciler/reconciler.go

```diff
@@ -749,6 +546,75 @@ func (r *Reconciler) pruneBalances(ctx context.Context, change *parser.BalanceCh
 	)
 }
 
+// skipAndPrune calls the ReconciliationSkipped
+// handler and attempts to prune.
+func (r *Reconciler) skipAndPrune(
+	ctx context.Context,
+	change *parser.BalanceChange,
+	skipCause string,
+) error {
+	if err := r.handler.ReconciliationSkipped(
+		ctx,
+		ActiveReconciliation,
+		change.Account,
+		change.Currency,
+		skipCause,
+	); err != nil {
+		return err
+	}
+
+	return r.updateQueueMapAndPrune(ctx, change)
+}
+
+// updateQueueMapAndPrune removes a *parser.BalanceChange
+// from the queueMap and attempts to prune the associated
+// *types.AccountCurrency's balances, if appropriate.
+func (r *Reconciler) updateQueueMapAndPrune(
+	ctx context.Context,
+	change *parser.BalanceChange,
+) error {
+	key := types.Hash(&types.AccountCurrency{
+		Account:  change.Account,
+		Currency: change.Currency,
+	})
+
+	r.queueMapMutex.Lock()
+	r.queueMap[key][change.Block.Index]--
+	if r.queueMap[key][change.Block.Index] > 0 {
+		r.queueMapMutex.Unlock()
+		return nil
+	}
+
+	// Cleanup indexes when we don't need them anymore
+	delete(r.queueMap[key], change.Block.Index)
+
+	// Sort indexes
+	indexes := []int64{}
+	for k := range r.queueMap[key] {
+		indexes = append(indexes, k)
+	}
+	sort.Slice(indexes, func(i, j int) bool { return indexes[i] < indexes[j] })
+
+	// Don't prune if there are indexes for this AccountCurrency
+	// less than this change.
+	if len(indexes) > 0 && change.Block.Index >= indexes[0] {
+		r.queueMapMutex.Unlock()
+		return nil
+	}
```

Comment:
Looks like we only care about the `len(indexes)` and `indexes[0]`. We should be able to achieve this without sorting? Not sure about the typical size of the `indexes` but this could be a difference between `O(nlogn)` vs `O(n)` 😛

Reply:
The size of indexes should be `~= the # of concurrent active reconciliations for a *types.AccountCurrency`. In the worst case, this could be the number of total items in the queue (if the *types.AccountCurrency` is active in each block and reconciliation is super slow).

Let me take another pass at a better algo here!

Reply:
Alright, I added a BST to make this much faster in the worst case. Shouldn't make much of a difference when there are a handful of indexes for an account but will make a big difference when there are a lot.

Reply:
Not sure if BST is the right call as it worsens the insertion from O(1) to O(logn) 

What about the following? O(n) to getMin() comparing to O(nlogn) with sort

```
min := func(a, b int) int {
	if a < b {
		return a
	}
	return b
}

minIndex := math.MaxInt64
for k := range r.queueMap[key] {
	minIndex = min(minIndex, k)
}

// Don't prune if there are indexes for this AccountCurrency
// less than this change.
if minIndex != math.MaxInt64 && change.Block.Index >= min {
	r.queueMapMutex.Unlock()
	return nil
}
```